gnmidiff/setrequest.go

```diff
@@ -267,52 +271,173 @@ func minimalSetRequestIntent(req *gpb.SetRequest) (setRequestIntent, error) {
 	return intent, nil
 }
 
+// binaryBase64 takes an input byte slice and returns it as a base64
+// encoded string.
+func binaryBase64(i []byte) string {
+	var b bytes.Buffer
+	encoder := base64.NewEncoder(base64.StdEncoding, &b)
+	encoder.Write(i)
+	encoder.Close()
+	return b.String()
+}
+
+// protoLeafToJSON converts a gNMI proto scalar to the equivalent RFC7951 JSON
+// representation.
+func protoLeafToJSON(tv *gpb.TypedValue) (interface{}, error) {
+	switch tv.GetValue().(type) {
+	case *gpb.TypedValue_StringVal:
+		return tv.GetStringVal(), nil
+	case *gpb.TypedValue_BoolVal:
+		return tv.GetBoolVal(), nil
+	case *gpb.TypedValue_IntVal:
+		return float64(tv.GetIntVal()), nil
+	case *gpb.TypedValue_UintVal:
+		return float64(tv.GetUintVal()), nil
+	case *gpb.TypedValue_DoubleVal:
+		return float64(tv.GetDoubleVal()), nil
+	case *gpb.TypedValue_LeaflistVal:
+		elems := tv.GetLeaflistVal().GetElement()
+		ss := make([]interface{}, len(elems))
+		for x, e := range elems {
+			var err error
+			if ss[x], err = protoLeafToJSON(e); err != nil {
+				return nil, err
+			}
+		}
+		return ss, nil
+	case *gpb.TypedValue_BytesVal:
+		return binaryBase64(tv.GetBytesVal()), nil
+	default:
+		return nil, fmt.Errorf("gnmidiff: TypedValue type %T is not a scalar type", tv.GetValue())
+	}
+}
+
 // populateUpdates populates all leaf updates at the given path into the intent.
 //
 // Note: The input path must NOT end with "/".
 //
 // e.g. /a for b/c="foo" would introduce an update of /a/b/c="foo" into the intent.
-func populateUpdates(intent *setRequestIntent, path string, tv *gpb.TypedValue) error {
+func populateUpdates(intent *setRequestIntent, path string, tv *gpb.TypedValue, newSchemaFn func() (*ytypes.Schema, error)) error {
```

Comment:
Is there a reason you don't simply just pass in a `ytypes.Schema`?

Reply:
This is to start with a clean root object each time unmarshalling happens. Otherwise there needs to be some `reflect` code that resets the root object each time to avoid previous unmarshalling done in previous invocations of `populateUpdates` polluting the results. Now you could argue why not just unmarshal everything and then send the updates. This would be valid aside from the corner case with list keys being populated, which I described in the new comment block I added.

Added comment explaining this.

Reply:
Done. This improved performance by a lot.